Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,352 @@
+def take (i : Nat) (v : Vector α n) : Vector α (min i n) :=
+  extract v 0 i
+
+/-- Vector of length `n` from a function on `Fin n`. -/
+def ofFn (f : Fin n → α) : Vector α n := ⟨Array.ofFn f, by {simp only [size_ofFn]}⟩
+
+/-- Create a vector from another with a provably equal length. -/
+protected def cast {n m : Nat} (h : n = m) : Vector α n → Vector α m
+  | ⟨x, p⟩ => ⟨x, h ▸ p⟩
+
+
+-- Swap functions
+/--
+Swaps two entries in an array, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference
+count of 1 when called.
+-/
+
+def swap (v : Vector α n) (i j : Fin n) : Vector α n :=
+  ⟨Array.swap v.toArray (v.size_eq.symm ▸ i) (v.size_eq.symm ▸ j), proof⟩
+  where
+    proof := by
+      simp only [size_swap]
+      exact v.size_eq
+      done
+
+/--
+Swaps two entries in an vector, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swap! (v : Vector α n) (i j : Nat) : Vector α n :=
+  ⟨Array.swap! v.toArray i j, by {simp only [size_swap!]; exact v.size_eq}⟩
+
+/--
+Swaps the entry with index `i : Fin n` in the vector for a new entry.
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt (v : Vector α n) (i : Fin n) (x : α) : α × Vector α n:=
+  let e := Vector.get v i
+  let v := Vector.set v i x
+  (e, v)
+
+instance [Inhabited α] : Inhabited (Vector α n) where
+  default := mkVector n default
+
+/--
+`swapAt! v i x` Swaps out the entry at index `i : Nat` in the vector for `x`, if the index is valid.
+Otherwise it panics
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt! (v : Vector α n) (i : Nat) (x : α) : α × Vector α n :=
+  if h : i < n then
+    swapAt v ⟨i, h⟩ x
+  else
+    have : Inhabited α := ⟨x⟩
+    panic! ("index " ++ toString i ++ " out of bounds")
+
+/--`range n` Returns a vector `#v[1,2,3,...,n-1]`-/
+def range (n : Nat) : Vector Nat n :=
+  ⟨Array.range n, by {simp}⟩
+
+/--
+Returns a vector of size `1` with a single element `v`
+-/
+def singleton (v : α) : Vector α 1 :=
+  mkVector 1 v
+
+/--
+Vector lookup function that takes an index `i` of type `USize`
+-/
+def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α :=
+  Array.uget v.toArray i (v.size_eq.symm ▸ h)
+
+instance : GetElem (Vector α n) Nat α fun _ i => i < n where
+  getElem := fun x i h => get x ⟨i, h⟩
+
+/--
+`pop v` returns the vector with the last element removed
+-/
+def pop (v : Vector α n) : Vector α (n-1) :=
+  ⟨Array.pop v.toArray, by {simp only [size_pop, v.size_eq]}⟩
+
+/--
+`shrink v m` removess the last `m` elements of a vector.
+-/
+def shrink (v : Vector α n) (m : Nat) : Vector α (n - m) :=
```

Comment:
A last one: this should use `Array.shrink`.

Reply:
The size proof for that gets stuck on shrink.loop. Also, is this really optimised for Arrays? There is no attribute that indicates this : https://github.com/leanprover/lean4/blob/367b97885a484a2e17e4cdade8dcd7fe05dbfb4c/src/Init/Data/Array/Basic.lean#L136

Reply:
Add this at the end of Batteries.Data.Array.Lemmas:
```lean
/-! ### shrink -/

theorem size_shrink_loop (a : Array α) (n) : (shrink.loop n a).size = a.size - n := by
  induction n generalizing a with simp [shrink.loop]
  | succ n ih => simp [ih]; omega

theorem size_shrink (a : Array α) (n) : (a.shrink n).size = min a.size n := by
  simp [shrink, size_shrink_loop]; omega
```

Reply:
It probably won't be optimized until the new compiler arrives but it is definitely on the todo list.

Reply:
Done.